Sources/Moya/Endpoint.swift

```diff
@@ -77,32 +79,35 @@ extension Endpoint {
             request.httpBody = data
             return request
         case let .requestParameters(parameters, parameterEncoding):
-            return try? parameterEncoding.encode(request, with: parameters)
+            return try parameterEncoding.encode(request, with: parameters)
         case let .uploadCompositeMultipart(_, urlParameters):
-            return try? URLEncoding(destination: .queryString).encode(request, with: urlParameters)
+            return try URLEncoding(destination: .queryString).encode(request, with: urlParameters)
         case let .downloadParameters(parameters, parameterEncoding, _):
-            return try? parameterEncoding.encode(request, with: parameters)
+            return try parameterEncoding.encode(request, with: parameters)
         case let .requestCompositeData(bodyData: bodyData, urlParameters: urlParameters):
             request.httpBody = bodyData
-            return try? URLEncoding(destination: .queryString).encode(request, with: urlParameters)
+            return try URLEncoding(destination: .queryString).encode(request, with: urlParameters)
         case let .requestCompositeParameters(bodyParameters: bodyParameters, bodyEncoding: bodyParameterEncoding, urlParameters: urlParameters):
             if bodyParameterEncoding is URLEncoding { fatalError("URLEncoding is disallowed as bodyEncoding.") }
-            guard let bodyfulRequest = try? bodyParameterEncoding.encode(request, with: bodyParameters) else { return nil }
-            return try? URLEncoding(destination: .queryString).encode(bodyfulRequest, with: urlParameters)
+            let bodyfulRequest = try bodyParameterEncoding.encode(request, with: bodyParameters)
+            return try URLEncoding(destination: .queryString).encode(bodyfulRequest, with: urlParameters)
         }
     }
 }
 
 /// Required for using `Endpoint` as a key type in a `Dictionary`.
 extension Endpoint: Equatable, Hashable {
     public var hashValue: Int {
-        return urlRequest?.hashValue ?? url.hashValue
+        let request = try? urlRequest()
+        return request?.hashValue ?? url.hashValue
     }
 
     public static func == <T>(lhs: Endpoint<T>, rhs: Endpoint<T>) -> Bool {
-        if lhs.urlRequest != nil, rhs.urlRequest == nil { return false }
-        if lhs.urlRequest == nil, rhs.urlRequest != nil { return false }
-        if lhs.urlRequest == nil, rhs.urlRequest == nil { return lhs.hashValue == rhs.hashValue }
-        return (lhs.urlRequest == rhs.urlRequest)
+        let lhsRequest = try? lhs.urlRequest()
+        let rhsRequest = try? rhs.urlRequest()
+        if lhsRequest != nil, rhsRequest == nil { return false }
+        if lhsRequest == nil, rhsRequest != nil { return false }
+        if lhsRequest == nil, rhsRequest == nil { return lhs.hashValue == rhs.hashValue }
```

Comment:
This is totally fine, though I will point out that if two different Endpoints both fail to produce a `urlRequest`, then they would be "equal". I'm sure that's fine, especially since it's consistent with `hashValue`. The only consequences would be using an `Endpoint` as key in a dictionary, or in a `Set`. I think it's worth a comment, but otherwise this is totally 👍

Reply:
If two different Endpoints both fail to produce a `urlRequest` it will compare both endpoints using the hashValue of the `url: String` property. 

So if two different Endpoints both fail to produce a `urlRequest` and they have the same `String` for the `url` -- they will be considered equal. I'll add a comment noting this behavior.